### PR TITLE
Book: Draft for exit code chapter

### LINKED ISSUE #39: Tracking Issue: Documenting Exit Codes
Being able to document exit codes would be a useful for many programs, especially when reimplementing existing programs where exit codes can have special meanings.

This thread serves to gather resources about discussions happening around this topic, with the goal that eventually we'll be in a position where it's possible to document all exit codes from our programs.

## Related Issues
- signal handling https://github.com/rust-lang-nursery/cli-wg/issues/27
- man pages https://github.com/rust-lang-nursery/cli-wg/issues/38
- error handling in CLI apps https://github.com/rust-lang-nursery/cli-wg/issues/12
- support development of well-behaved CLI apps https://github.com/rust-lang-nursery/cli-wg/issues/21

## Links
- https://internals.rust-lang.org/t/mini-pre-rfc-redesigning-process-exitstatus/5426/15
- https://github.com/rust-lang/rust/issues/48711

### PR DESCRIPTION
**Pull request by killercup**
Resolves #39

### CODE REVIEW

File: src/in-depth/exit-code.md [hunk 1]
```diff
@@ -1 +1,55 @@
...
+In addition to
+[telling the user about errors](human-communication.md),
+on most systems,
+when a process exits,
+it also emits an exit code
+(an integer between 0 and 255).
+You should try to emit the correct code
+for your program's state.
+For example,
+in the ideal case when your program succeeded,
```
Comment (Dylan-DPC) [inline src/in-depth/exit-code.md]: when your program succeeds,

File: src/in-depth/exit-code.md [hunk 2]
```diff
@@ -1 +1,55 @@
...
+it should exit with `0`.
+
+When an error ocurred, it gets a bit more complicated, though.
```
Comment (Dylan-DPC) [inline src/in-depth/exit-code.md]: s/occured/occurs

File: src/in-depth/exit-code.md [hunk 3]
```diff
@@ -1 +1,55 @@
...
+In the wild,
+a lot of tools exit with `1` when a general failure ocurred.
+Currently, Rust set and exit code of `101` when the process panicked.
```
Comment (Dylan-DPC) [inline src/in-depth/exit-code.md]: Currently, Rust sets an exit code of ...

File: src/in-depth/exit-code.md [hunk 4]
```diff
@@ -1 +1,55 @@
...
+Beyond that, many people have done many things in their programs.
```
Comment (Dylan-DPC) [inline src/in-depth/exit-code.md]: Beyond that, ~many~ people have ...

File: src/in-depth/exit-code.md [hunk 5]
```diff
@@ -1 +1,55 @@
...
+
+So, what to do?
+The BSD ecosystem has collected a common definition for their exit codes
+in a system-provided header file called [`sysexits.h`]
+The Rust library [`exitcode`] provides these same codes
```
Comment (Dylan-DPC) [inline src/in-depth/exit-code.md]: provides the same codes which are ready ...

File: src/in-depth/exit-code.md [hunk 6]
```diff
@@ -1 +1,55 @@
...
+ready to be used in your application.
+Please see it's API documentation for the possible values to use.
```
Comment (Dylan-DPC) [inline src/in-depth/exit-code.md]: s/it's/its

**Comment by killercup** [comment]
Thanks so much, @Dylan-DPC!

File: src/in-depth/exit-code.md [hunk 7]
```diff
@@ -1 +1,55 @@
...
+
+A program doesn't always succeed.
+And when an error occurs,
+you should make sure to emit the necessary information correctly.
+In addition to
+[telling the user about errors](human-communication.md),
+on most systems,
+when a process exits,
+it also emits an exit code
+(an integer between 0 and 255).
```
Comment (Screwtapello) [inline src/in-depth/exit-code.md]: On POSIX, an exit code is a u8, while on Win32 an exit code is a u32 (weirdly, `std::process::exit` takes an i32, I'm not sure why). It's probably best not to mention the actual limits until you get to a "platform-specific behaviour" section.

File: src/in-depth/exit-code.md [hunk 8]
```diff
@@ -1 +1,55 @@
...
+
+When an error occurs, it gets a bit more complicated, though.
+In the wild,
+a lot of tools exit with `1` when a general failure ocurred.
+Currently, Rust set and exit code of `101` when the process panicked.
+Beyond that, people have done many things in their programs.
+
+So, what to do?
+The BSD ecosystem has collected a common definition for their exit codes
+in a system-provided header file called [`sysexits.h`]
```
Comment (Screwtapello) [inline src/in-depth/exit-code.md]: Is it appropriate to assume the reader will understand "a system-provided header file"? A lot of people certainly will, but somebody who comes here right after finishing TRPL might not.

If this book is aimed at an audience that already knows what header files are, that's fine, I just thought I'd mention it.

File: src/in-depth/exit-code.md (hunk 7 above)
Comment (killercup) [reply src/in-depth/exit-code.md]: How about "an integer between 0 and 255 is compatible with most platforms"?

File: src/in-depth/exit-code.md (hunk 8 above)
Comment (killercup) [reply src/in-depth/exit-code.md]: Good point! I wanted to point to the source of the exit code list but you're right that is probably more confusing than helpful.

**Comment by killercup** [comment]
Merging to get this on the site, but always re-reviewable in #58 :)

**Pull request merged**